tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java

```diff
@@ -0,0 +1,208 @@
+import com.pingcap.tikv.expression.ColumnRef;
+import com.pingcap.tikv.expression.ComparisonBinaryExpression;
+import com.pingcap.tikv.expression.ComparisonBinaryExpression.NormalizedPredicate;
+import com.pingcap.tikv.expression.Constant;
+import com.pingcap.tikv.expression.Expression;
+import com.pingcap.tikv.expression.IsNull;
+import com.pingcap.tikv.expression.LogicalBinaryExpression;
+import com.pingcap.tikv.expression.Not;
+import com.pingcap.tikv.expression.visitor.CanBePrunedValidator.ContextPartExpr;
+import com.pingcap.tikv.meta.TiPartitionDef;
+import com.pingcap.tikv.meta.TiPartitionInfo;
+import com.pingcap.tikv.meta.TiPartitionInfo.PartitionType;
+import com.pingcap.tikv.meta.TiTableInfo;
+import com.pingcap.tikv.predicates.PredicateUtils;
+import java.util.ArrayList;
+import java.util.List;
+
+public class PrunedPartitionBuilder extends DefaultVisitor<RangeSet<Long>, Void> {
+
+  private static Expression partExpr;
+
+  private static void throwOnError(Expression node) {
+    final String errorFormat = "Unsupported conversion to Range: %s";
+    throw new TiExpressionException(String.format(errorFormat, node));
+  }
+
+  protected RangeSet<Long> process(Expression node, Void context) {
+    throwOnError(node);
+    return null;
+  }
+
+  /**
+   * return false if table cannot be pruning or partition table is not enabled. return true if
+   * partition expression is simple.
+   *
+   * @param tblInfo is table info which is used for creaeting columnRef.
+   * @param filter is a where condition. It must be a cnf and does not contain not or isnull.
+   * @return true if partition pruning can apply on filter.
+   */
+  private static boolean canBePruned(TiTableInfo tblInfo, Expression filter) {
+    if (!tblInfo.isPartitionEnabled()) {
+      return false;
+    }
+
+    if (TiPartitionInfo.toPartType((int) tblInfo.getPartitionInfo().getType())
+        != PartitionType.RangePartition) {
+      return false;
+    }
+
+    // if query is select * from t, then filter will be null.
+    if (filter == null) return false;
+
+    // Ideally, part expr should be parsed by a parser
+    // but it requires a lot work, we chose a brute force way
+    // to workaround.
+    boolean canBePruned = true;
+    partExpr = ColumnRef.create(tblInfo.getPartitionInfo().getExpr(), tblInfo);
+    if (!tblInfo.getPartitionInfo().getExpr().contains("(")) {
+      // skip not(isnull(col)) case
+      canBePruned = CanBePrunedValidator.canBePruned(filter, new ContextPartExpr(partExpr));
+    }
+
+    return canBePruned;
+  }
+
+  private static List<Expression> filterFilters(List<Expression> filters) {
+    List<Expression> filteredFilters = new ArrayList<>();
+    for (Expression expr : filters) {
+      if (expr instanceof Not || expr instanceof IsNull) {
+        continue;
+      }
+      filteredFilters.add(expr);
+    }
+    return filteredFilters;
+  }
+
+  @Override
+  @SuppressWarnings("Duplicates")
```

Comment:
Is it possible to reuse index range builder code?

Reply:
It is possible. I will work on that.